src/net/sourceforge/kolmafia/session/YouRobotManager.java

```diff
@@ -16,92 +22,528 @@
+    // all the various sets and maps from the constructors.
+    RobotUpgrade[] values = RobotUpgrade.values();
+  }
+
+  private static RobotUpgrade urlFieldsToUpgrade(Part part, String chosenPart) {
+    if (part == Part.CPU) {
+      return keywordToCPU.get(chosenPart);
+    }
+    return partToIndexMap.get(part).get(StringUtilities.parseInt(chosenPart));
+  }
+
+  // *** Current state of Configuration
+
+  private static final Map<Part, RobotUpgrade> currentParts = new HashMap<>();
+  private static final Set<RobotUpgrade> currentCPU = new HashSet<>();
+
+  // For testing
+  public static void reset() {
+    currentParts.clear();
+    currentCPU.clear();
+  }
+
+  public static boolean hasEquipped(String name) {
+    RobotUpgrade upgrade = nameToUpgrade.get(name);
+    return (upgrade == null) ? false : hasEquipped(upgrade);
+  }
+
+  public static boolean hasEquipped(RobotUpgrade upgrade) {
+    Part part = upgrade.getPart();
+    switch (part) {
+      case TOP:
+      case LEFT:
+      case RIGHT:
+      case BOTTOM:
+        return currentParts.get(part) == upgrade;
+      case CPU:
+        return currentCPU.contains(upgrade);
+    }
+    return false;
+  }
+
+  // *** Public methods to parse robot info from KoL responses
+
+  // Parse avatar from:
+  //     CharSheetRequest
+  //     CharPaneRequest
+  //     ChoiceManager (Reassembly Station)
 
   private static final Pattern AVATAR =
       Pattern.compile("(otherimages/robot/(left|right|top|bottom|body)(\\d+).png)\"");
 
   public static void parseAvatar(final String text) {
+    Map<Part, RobotUpgrade> parts = new HashMap<>();
+
     List<String> images = new ArrayList<>();
     Matcher m = AVATAR.matcher(text);
     while (m.find()) {
       images.add(m.group(1));
       String section = m.group(2);
-      int config = StringUtilities.parseInt(m.group(3));
-      Preferences.setInteger("youRobot" + StringUtilities.toTitleCase(section), config);
+      int index = StringUtilities.parseInt(m.group(3));
+
+      if (!section.equals("body")) {
+        // Set the "current" configuration variables
+        Part part = keywordToPart.get(section);
+        RobotUpgrade upgrade = partToIndexMap.get(part).get(index);
+        parts.put(part, upgrade);
+      } else {
+        // This never changes, but make sure it is set once, at least
+        Preferences.setInteger("youRobotBody", index);
+      }
+    }
+
+    boolean changed = false;
+    for (Entry<Part, RobotUpgrade> entry : parts.entrySet()) {
+      Part part = entry.getKey();
+      RobotUpgrade previous = currentParts.get(part);
+      RobotUpgrade upgrade = entry.getValue();
+      if (upgrade != previous) {
+        // Add to current configuration
+        currentParts.put(part, upgrade);
+        // Set the legacy properties for use by scripts
+        Preferences.setInteger("youRobot" + part.getSection(), upgrade.getIndex());
+        // *** Fire listeners?
```

Comment:
presumably these are both TODO? Combat skills has a disabled test.

Reply:
Right. I am about to post my list of TODOs. Firing listeners is pointless unless GUI elements are listening for them. :)